onnx/hub.py

```diff
@@ -363,12 +363,35 @@ def download_model_with_test_data(
                 "download the model from the model hub."
             )
 
+    def members_filter(tarfile, base):
+        result = []
+        for member in tarfile:
+            member_path = os.path.join(base, member.name)
+            abs_base = os.path.abspath(base)
+            abs_member = os.path.abspath(member_path)
+            if not abs_member.startswith(abs_base):
+                raise AssertionError(
+                    "The tarball file in downloading model contains directory traversal sequence which may contain harmful payload."
+                )
+            elif member.issym() or member.islnk():
+                raise AssertionError(
+                    "The tarball file in downloading model contains symbolic links which may contain harmful payload."
+                )
+            result.append(member)
+        return result
+
     with tarfile.open(local_model_with_data_path) as model_with_data_zipped:
         # FIXME: Avoid index manipulation with magic numbers
         local_model_with_data_dir_path = local_model_with_data_path[
             0 : len(local_model_with_data_path) - 7
         ]
-        model_with_data_zipped.extractall(local_model_with_data_dir_path)
+        # FIX tarball directory traversal
+        model_with_data_zipped.extractall(
+            path=local_model_with_data_dir_path,
+            members=members_filter(
+                model_with_data_zipped, local_model_with_data_dir_path
```

Comment:
Looks like there is a filter option available since Python 3.12. We should conditionally enable this option by checking `hasattr(tarfile, 'data_filter')`. Ref: https://docs.python.org/3/library/tarfile.html#supporting-older-python-versions

Reply:
Added python3.12 data_filter